### ISSUE TITLE
Implement Search by Symbol Across a Workspace

### ISSUE THREAD

**Issue by j-potts** [scalameta/metals-feature-requests#33]
**Is your feature request related to a problem? Please describe.**
Search by symbol across a workspace: https://code.visualstudio.com/docs/editor/editingevolved#_open-symbol-by-name

**Describe the solution you'd like**
Similar to the current search by symbol within open documents, it should be possible to search across a workspace with results including the [symbol kind](https://microsoft.github.io/language-server-protocol/specification#textDocument_documentSymbol).

**Describe alternatives you've considered**
I've used [Stags](https://github.com/pjrt/stags) with limited success, but it does not offer the specificity of something like symbol kind. Otherwise, I tend to just use the editors built-in search/find, or open a file and search by symbol within the file.

**Comment by matthughes** [comment]
You could use the same functionality as with search by type.  By default, the symbol view only shows symbols in the current document.  However, if you append `;` we could search the entire codebase.  Or something more arcane where `;` means current source code and `;;` means current source code + libraries on classpath.

**Comment by olafurpg** [comment]
Thank you for reporting! Can you please elaborate on this feature request?

Metals already supports "workspace symbol" for locally defined symbols as well as symbols from the library dependency classpath. However, currently, this functionality is limited to classes, traits and objects. It's not possible to search for methods or variables, but it might be possible with a bit of thinking

**Comment by j-potts** [comment]
Thanks for getting in touch @olafurpg ! This feature would work similar to the existing symbol look-up by name, but instead of being bound to a local scope, it would be able to lookup across a workspace. For example: https://code.visualstudio.com/docs/editor/editingevolved#_open-symbol-by-name

This would include methods, variables, traits, classes, etc. Does that help?

**Comment by olafurpg** [comment]
Have you tried “workspace symbol”? This feature is already supported in Metals but it’s limited to classes, traits and objects

**Comment by j-potts** [comment]
Yes, but because a file that doesn't compile won't allow me to click-through on anything, I find myself searching method names a lot. If they were available via the symbol search, it would be much quicker for me. It's possible I'm just missing some feature in my workflow, though.

**Comment by olafurpg** [comment]
Navigation should work mostly OK even if a file doesn't currently compile, as long as it compiled at least once before. Navigation should also work for things you have just typed (which was not the case before)

**Comment by tgodzik** [comment]
@olafurpg how hard would it be to implement it also for methods? We could enable it via a setting, so it wouldn't add any overhead if someone doesn't need it.

**Comment by olafurpg** [comment]
I don't think it would be too hard to support searching for workspace-defined methods and I think we can do it with little overhead. We might also be able to index classpath-defined methods that are already referenced in the workspace with low overhead since we already walk through the SemanticDB files. These two combined would probably address most common use-cases for searching methods.

However, I'm on the fence about eagerly indexing the entire classpath to search for classpath-defined methods which are not referenced in the workspace. I estimate that would introduce significant indexing overhead for questionable gain.

If we do this it should be enabled by default and we should make sure it doesn't introduce a regression. If the feature is not good enough to be on by default then we probably don't want to support it.

**Comment by j-potts** [comment]
I agree, eagerly indexing methods that are not referenced within the workspace would be wasteful. Maybe a separate path/call to index non-referenced methods could be triggered by user input.

**Comment by olafurpg** [comment]
I would start by indexing only workspace defined and referenced symbols to see how far that goes. We can discuss indexing other symbols separately.

**Comment by nkgm** [comment]
This would be a huge productivity booster. Any updates?

**Comment by tgodzik** [comment]
Nothing currently, but maybe we could do it at least partially with https://github.com/scalameta/metals/issues/1381 limited to workspace for now :thinking:

**Comment by tgodzik** [comment]
This is now supported within the workspace, but we don't calculate methdo/value symbols for classpath symbols. We could improve it to include things referenced in the workspace  in the future.

**Comment by tgodzik** [comment]
Actually, that is exactly what we have in the title. Will close it for now, let me know if anyone feels it would be good to also search the dependencies, but that in reality might be way too much information.
